src/solver/mod.rs

```diff
@@ -579,45 +629,55 @@ impl<D: DependencyProvider, RT: AsyncRuntime> Solver<D, RT> {
     ///    implementation of this step.
     ///
     /// The solver loop can be found in [`Solver::resolve_dependencies`].
-    fn run_sat(&mut self) -> Result<(), UnsolvableOrCancelled> {
-        assert!(self.decision_tracker.is_empty());
-        let mut level = 0;
+    ///
+    /// Returns `Ok(true)` if a solution was found for `solvable`. If a solution was not
+    /// found, returns `Ok(false)` if some decisions have already been made by the solver
+    /// (i.e. the decision tracker stack is not empty). Otherwise, returns `Err` on no solution.
+    fn run_sat(&mut self, solvable: InternalSolvableId) -> Result<bool, UnsolvableOrCancelled> {
```

Comment:
I think the return type doesnt quite work. I would expect the `Cancelled` return value to still be returned when a cancellation is triggered even for the additional solvables. Why not just always return `UnsolvableOrCancelled` and catch the unsolvable error in `solve_with_additional`.

Reply:
`Err(Cancelled)` should still be returned - the only change is returning `Ok(false)` instead of `Err(Unsolvable)` for unsolvable additional solvables. I did not go for the approach you recommended because it would lead to calling `analyze_unsolvable` being unnecessarily called (though I suppose the compiler may be able to optimize it away). `analyze_unsolvable` also currently seems to assume (via a `debug_assert_eq` statement ) that the highest level of decisions made is 1, and I didn't want to break the assumption (the level would be higher than 1 for unsolvable additional solvables).

Reply:
👍 Maybe clarify that in the comment.

Reply:
Updated the doc comment to specify that it still returns `Err(Cancelled)`. Did you also want me to include an explanation as to why I don't return `Unsolvable` for additional solvables?

Reply:
Nope this is fine! Thank you :)